notebooks/Analyze_user_behavior_across_different_lending_protocols.ipynb

```diff
@@ -196,90 +217,92 @@
-    "\n",
-    "liquidity_data = df_loans[df_loans['Collateral (USD)'] > 0]\n",
-    "\n",
-    "# Initialize a dictionary to store users and their associated protocols for liquidity\n",
-    "user_protocols_liquidity = defaultdict(set)\n",
-    "\n",
-    "# Populate the dictionary\n",
-    "for _, row in liquidity_data.iterrows():\n",
-    "    user = row['User']\n",
-    "    protocol = row['Protocol']\n",
-    "    user_protocols_liquidity[user].add(protocol)\n",
-    "\n",
     "# Count the number of protocols each user lends on\n",
-    "user_protocol_counts_liquidity = Counter([len(protocols) for protocols in user_protocols_liquidity.values()])\n",
+    "liquidity_data = df_loans[df_loans['Collateral (USD)'] > 0]\n",
+    "protocol_count_series = liquidity_data.groupby('User')['Protocol'].count().value_counts()\n",
     "\n",
-    "# Convert the counter to a DataFrame for better readability\n",
-    "protocol_count_df_liquidity = pd.DataFrame.from_dict(user_protocol_counts_liquidity, orient='index').reset_index()\n",
+    "# Convert the series to a DataFrame for better readability\n",
+    "protocol_count_df_liquidity = protocol_count_series.reset_index()\n",
     "protocol_count_df_liquidity.columns = ['Number of Protocols', 'Number of Users']\n",
-    "\n",
     "# Sort the DataFrame by the number of protocols\n",
     "protocol_count_df_liquidity = protocol_count_df_liquidity.sort_values(by='Number of Protocols')"
    ]
   },
+  {
+   "cell_type": "code",
+   "execution_count": 6,
+   "id": "86651f0f-a763-4780-83ac-65fcc9edce9c",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "   Number of Protocols  Number of Users\n",
+      "0                    1           402959\n",
+      "1                    2            71145\n",
+      "2                    3             4510\n",
+      "3                    4               19\n",
+      "4                    5                1\n"
+     ]
+    }
+   ],
+   "source": [
+    "print(protocol_count_df_liquidity)"
+   ]
+  },
   {
    "cell_type": "markdown",
    "id": "db697ca0-30fe-4d21-a5fc-b01c5db05ad9",
    "metadata": {},
    "source": [
-    "##### Users Providing Liquidity Across the Top 3 Protocols"
+    "### Users Providing Liquidity Across the Top 3 Protocols"
    ]
   },
   {
    "cell_type": "code",
-   "execution_count": 11,
+   "execution_count": 7,
    "id": "2853f77c-143b-4d6a-b584-20a515fa7d09",
    "metadata": {},
    "outputs": [],
    "source": [
     "## Helper funcitons:\n",
     "# Function to get unique users per protocol\n",
-    "def get_unique_users(df, value_column):\n",
-    "    protocol_users = defaultdict(set)\n",
-    "    for protocol in df['Protocol'].unique():\n",
-    "        users = set(df[df['Protocol'] == protocol]['User'])\n",
-    "        protocol_users[protocol].update(users)\n",
+    "def get_unique_users(df:pandas.DataFrame, value_column:str) -> Dict[str, Set[str]]:\n",
```

Comment:
Isn't the value column always set to `'User'`? It's also in the name of the function, it would be weird to call it on another column.

The alternative is to rename to function to `get_unique_values`.